BlueNRG: allow setting public address

arbrauns opened the pull request:
The zephyr bluetooth stack expects the controller to know its public address, if any. At least for BlueNRG, the public address is forgotten with every reset/power cycle, so there needs to be a way to set it from within zephyr. This is accomplished using the `Aci_Hal_Write_Config_Data` HCI command, as described in [PM0237].

[PM0237]: https://www.st.com/resource/en/programming_manual/pm0237-bluenrg-bluenrgms-stacks-programming-guidelines-stmicroelectronics.pdf

arbrauns committed:
drivers: refactor BlueNRG SPI driver for more config options

"LL only" is not the only config option of potential interest, e.g. the
public address is also important.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns committed:
drivers: allow setting BlueNRG public address

The zephyr bluetooth stack expects the controller to know its public
address, if any. At least for BlueNRG, the public address is forgotten with
every reset/power cycle, so there needs to be a way to set it from within
zephyr. This is accomplished using the `Aci_Hal_Write_Config_Data` HCI
command, as described in PM0237.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

jhedberg commented on include/zephyr/drivers/bluetooth/hci_driver.h:
```diff
@@ -296,6 +296,19 @@ struct net_buf *bt_hci_cmd_complete_create(uint16_t op, uint8_t plen);
  */
 struct net_buf *bt_hci_cmd_status_create(uint16_t op, uint8_t status);
 
+#if defined(CONFIG_BT_BLUENRG_ACI)
+/** Set the BlueNRG public address.
+ *
+ * This function ensures the BlueNRG is configured using the specified public address. It needs to
+ * be called before bt_enable().
+ *
+ * @param addr     Public address to use.
+ *
+ * @return 0 on success, negative errno otherwise
+ */
+int bt_spi_bluenrg_set_pubaddr(bt_addr_t *addr);
+#endif /* CONFIG_BT_BLUENRG_ACI */
```
`hci_driver.h` is not really the right place to put vendor/driver specific APIs.

This might be a good opportunity to add such a feature to Zephyr on a more generic level, by making it possible to pass a public address to the `bt_id_create()` API. Right now this API explicitly rejects anything that's not a static random address:

https://github.com/zephyrproject-rtos/zephyr/blob/2171f8b7a1770f5818fce0083175c8360004672c/subsys/bluetooth/host/id.c#L1284-L1289 

It's also worth noting that we've actually specified a Zephyr vendor HCI command that could be used for this, but so far I don't think any controller implements it (even Zephyr's own):
https://github.com/zephyrproject-rtos/zephyr/blob/2171f8b7a1770f5818fce0083175c8360004672c/doc/connectivity/bluetooth/api/hci.txt#L410-L415

What I'd propose for now, is to do the following:
 - Create a new HCI driver Kconfig option which adds a new callback to the HCI driver struct for setting the public address.
 - Update bt_id_create() to accept public addresses, and make it call the new driver callback (if available, otherwise return an error like right now)

jhedberg replied:
> * Update bt_id_create() to accept public addresses, and make it call the new driver callback (if available, otherwise return an error like right now)

Correction: you can't (and shouldn't) call the new "set public address" callback through `bt_id_create()` (that's only used to store the address), rather you need to modify the HCI init sequence to do this, most likely first thing in the `bt_init()` function.

arbrauns replied:
So where should I pass the public address from application to subsys? As an argument to `bt_enable()`->`init_work()`->`bt_init()`->`hci_init()`->some `struct bt_dev` method? That seems a bit roundabout.

jhedberg replied:
> So where should I pass the public address from application to subsys? As an argument to `bt_enable()`->`init_work()`->`bt_init()`->`hci_init()`->some `struct bt_dev` method? That seems a bit roundabout.

You would pass it to `bt_id_create()` before calling `bt_enable()`

jhedberg replied:
Just to clarify a bit more: `bt_id_create()` would store the address in `bt_dev.id_addr[0]` and the HCI init sequence would pick it from there (the init sequence would check for the address type and if it's public it would call the new HCI driver callback)

arbrauns replied:
Ahh I see, that makes sense, thanks for the explanation.

arbrauns replied:
Updated as requested.

arbrauns committed:
bluetooth: add HCI driver parameter to set controller's public address

This allows HCI drivers to expose vendor-specific functions to set the
public address.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

jhedberg commented on include/zephyr/drivers/bluetooth/hci_driver.h:
```diff
@@ -296,6 +314,19 @@ struct net_buf *bt_hci_cmd_complete_create(uint16_t op, uint8_t plen);
  */
 struct net_buf *bt_hci_cmd_status_create(uint16_t op, uint8_t status);
 
+#if defined(CONFIG_BT_BLUENRG_ACI)
+/** Set the BlueNRG public address.
+ *
+ * This function ensures the BlueNRG is configured using the specified public address. It needs to
+ * be called before bt_enable().
+ *
+ * @param addr     Public address to use.
+ *
+ * @return 0 on success, negative errno otherwise
+ */
+int bt_spi_bluenrg_set_pubaddr(bt_addr_t *addr);
```
I guess you forgot to remove this?

jhedberg commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -224,26 +231,39 @@ static bool exit_irq_high_loop(void)
 	return 0;
 }
+
+#if defined(CONFIG_BT_HCI_SET_PUBADDR)
+int bt_spi_bluenrg_set_pubaddr(bt_addr_t *addr)
```
Since this function is BlueNRG-specific, I guess what you want is to check for CONFIG_BT_BLUENRG_ACI instead of CONFIG_BT_HCI_SET_PUBADDR

arbrauns replied:
This is within an `#if defined(CONFIG_BT_BLUENRG_ACI)` guard.

jhedberg replied:
> This is within an `#if defined(CONFIG_BT_BLUENRG_ACI)` guard.

Ok, in that case it's redundant, since `BT_BLUENRG_ACI` selects `BT_HCI_SET_PUBADDR`, i.e. it's impossible to disable the option.

jhedberg commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -517,6 +537,9 @@ static const struct bt_hci_driver drv = {
 	.bus		= BT_HCI_DRIVER_BUS_SPI,
 #if defined(CONFIG_BT_BLUENRG_ACI)
 	.quirks		= BT_QUIRK_NO_RESET,
+#if defined(CONFIG_BT_HCI_SET_PUBADDR)
+	.set_public_address = bt_spi_bluenrg_set_pubaddr,
```
Same here

jhedberg commented on include/zephyr/drivers/bluetooth/hci_driver.h:
```diff
@@ -159,6 +159,24 @@ struct bt_hci_driver {
+	 *
+	 * @return 0 on success or negative error number on failure.
+	 */
+	int (*set_public_address)(bt_addr_t *addr);
```
Name seems generally fine, but maybe we want to be consistent with shortening "address" to "addr" like in other Bluetooth APIs (e.g. `bt_addr_t`)?

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4020,6 +4020,22 @@ int bt_enable(bt_ready_cb_t cb)
+	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
+#if defined(CONFIG_BT_HCI_SET_PUBADDR)
+		if (bt_dev.drv->set_public_address != NULL) {
+			err = bt_dev.drv->set_public_address(&bt_dev.id_addr[BT_ID_DEFAULT].a);
```
I was originally thinking that this would happen *after* the call to `drv->open()`, since usually this operation would require having the HCI transport available (which `open()` ensures), unless this requires the driver to temporarily store the address before actually writing it to the controller. Any reason why you think this absolutely needs to be before `open()`?

arbrauns replied:
I'm testing it right now, but I think BlueNRG may require setting this option before `open()` completes.

arbrauns replied:
Right after `drv->open()` is fine actually, just not after `common_init()`. I've put it right after `drv->setup()` now.

jhedberg replied:
> Right after `drv->open()` is fine actually, just not after `common_init()`. I've put it right after `drv->setup()` now.

Could we do it right before setup() instead, in case the setup routine wants to use this address? E.g. the `hci_psoc6_bless.c` HCI driver uses the `setup()` callback to program a public address to the controller:
https://github.com/zephyrproject-rtos/zephyr/blob/063ce9caf54fa656f02ae48f3c9d537659a10dec/drivers/bluetooth/hci/hci_psoc6_bless.c#L199-L208

One could imagine that someone might want to extend support in this driver to override the default public address with another one provided from the app, in which case one way to do this would be to use the `set_public_addr()` callback to store the alternate address and then program this instead of the default one in `setup()`.

arbrauns replied:
Makes sense, done.

msmttchr replied:
> I'm testing it right now, but I think BlueNRG may require setting this option before `open()` completes.

`bt_spi_open` will reset the BlueNRG device, so everything you do before is going to be lost.

arbrauns replied:
No, I had previously sent the ACI command "in" `open()` (in `bt_spi_handle_vendor_evt()`), which worked, but then moved it after `common_init()`, which didn't work anymore. It doesn't actually have to happen in `open()`, just needs to be before `common_init()`, not entirely sure why.

msmttchr replied:
OK, I think this is because that command needs to be run after reset, since it is disabled after standard command are received. I am quoting the programming manual:

> The command ACI_HAL_WRITE_CONFIG_DATA should be sent to BlueNRG and 
> BlueNRG-MS devices by the uC before starting BLE operations (after each power-up or 
> reset of BlueNRG).

I guess you run the command in the `common_init` after some other commands. On the other hand, I would have expected the error 0x0C (Command Disallowed) returned.

jhedberg commented on subsys/bluetooth/host/id.c:
```diff
@@ -1282,15 +1282,26 @@ int bt_id_create(bt_addr_le_t *addr, uint8_t *irk)
 			return -EALREADY;
 		}
+
+		if (addr->type == BT_ADDR_LE_PUBLIC && !IS_ENABLED(CONFIG_BT_PRIVACY)) {
```
Privacy doesn't really have anything to do with this. Even if privacy is enabled, the *identity address* can still be a public address. So that condition should be removed, I think.

That said, would be good to double-check that we don't have any code which assumes the identity address is always a static random one.

arbrauns replied:
I'm afraid I don't really understand how bluetooth privacy works then, I've removed the check though.

jhedberg replied:
> I'm afraid I don't really understand how bluetooth privacy works then, I've removed the check though.

Thanks. Devices always have an identity address, which can be either public or static random. Even with privacy enabled, the identity address gets shared with peers during pairing, together with the Identity Resolving Key.

jhedberg reviewed (changes requested):
Looks pretty good and pretty close to what I had in mind, thanks! See my inline comments for some things that still need tweaking.

jhedberg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -163,6 +163,15 @@ config BT_DRIVER_QUIRK_NO_AUTO_DLE
 	  This has to be enabled when the BLE controller connected is Zephyr
 	  open source controller.
 
+config BT_HCI_SET_PUBADDR
```
The only small issue I have here is the consistency with the naming. The new callback is called `set_public_addr`, so maybe it'd make sense for the Kconfig option to be called `BT_HCI_SET_PUBLIC_ADDR`. The other alternative would be to call both `set_pub_addr()` and `BT_HCI_SET_PUB_ADDR`.

arbrauns replied:
Renamed to `BT_HCI_SET_PUBLIC_ADDR`, it was bugging me too but not enough to actually do it ;)

jhedberg commented on subsys/bluetooth/host/id.c:
```diff
@@ -1281,20 +1281,31 @@ int bt_id_create(bt_addr_le_t *addr, uint8_t *irk)
+			/* set the single public address */
+			bt_addr_le_t *default_id = &bt_dev.id_addr[BT_ID_DEFAULT];
+
+			if (bt_dev.id_count > 0 && !bt_addr_le_eq(default_id, BT_ADDR_LE_ANY)) {
```
I don't think you need to check for `BT_ADDR_LE_ANY` here, since we don't allow creating identities with that address. Or to put it another way, if `id_count > 0` then the default id has to be some valid address rather than `BT_ADDR_LE_ANY`. If you look at `id.c` and the `id_create()` function, you'll see that if it gets `BT_ADDR_LE_ANY` it branches off to generate a static random address instead.

arbrauns replied:
It could theoretically have been deleted by `bt_id_delete()`, which sets it to `BT_ADDR_LE_ANY`, but only decrements `id_count` if it was the last id. Probably a very niche edge case, I can remove the check if it adds more confusion than it solves.

jhedberg replied:
Or is there other some use case I'm not seeing here?

jhedberg replied:
> It could theoretically have been deleted by `bt_id_delete()`, which sets it to `BT_ADDR_LE_ANY`, but only decrements `id_count` if it was the last id. Probably a very niche edge case, I can remove the check if it adds more confusion than it solves.

I think we should probably prohibit deleting a public address identity.

arbrauns replied:
Oh, actually that's already the case, `bt_id_delete()` forbids deleting the default ID. I've applied your suggestion of `(bt_dev.id_count != 0)`.

alwa-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3638,6 +3638,23 @@ static void hci_vs_init(void)
 {
 	int err;
+
+	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
+#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
+		if (bt_dev.drv->set_public_addr != NULL) {
+			err = bt_dev.drv->set_public_addr(&bt_dev.id_addr[BT_ID_DEFAULT].a);
+			if (err) {
+				LOG_ERR("Failed to set public address (%d)", err);
+				return err;
+			}
+		} else
+#endif /* defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR) */
+		{
+			LOG_ERR("Setting public address not supported by driver");
+			return -ENOTSUP;
+		}
+	}
```
This can be refactored so that the ifdefs "fit into the parse tree" and look like true ifs. It think that's much more preferable.

```suggestion
	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
		if (bt_dev.drv->set_public_addr != NULL) {
			err = bt_dev.drv->set_public_addr(&bt_dev.id_addr[BT_ID_DEFAULT].a);
			if (err) {
				LOG_ERR("Failed to set public address (%d)", err);
			}
			return err;
		}
#endif /* defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR) */

		LOG_ERR("Setting public address not supported by driver");
		return -ENOTSUP;
	}
```

I also have a preference for using the IF_ENABLED macro instead of the ifdefs in these cases, as this gets a more readable indentation.

```suggestion
	if (bt_dev.id_count > 0 && bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_PUBLIC) {
		IF_ENABLED(CONFIG_BT_HCI_SET_PUBLIC_ADDR, ({
			if (bt_dev.drv->set_public_addr != NULL) {
				err = bt_dev.drv->set_public_addr(&bt_dev.id_addr[BT_ID_DEFAULT].a);
				if (err) {
					LOG_ERR("Failed to set public address (%d)", err);
				}
				return err;
			}
		})

		LOG_ERR("Setting public address not supported by driver");
		return -ENOTSUP;
	}
```

arbrauns replied:
I don't think that's correct, to me it looks like that would always return after calling `set_public_addr()`.

carlescufi reviewed (changes requested):
Thanks for the patch. I would like to understand where would the application get the public address from? Typically a public address is stored in the controller, but in your case you seem to imply that it is on the application side, since you need to call `bt_id_create()` with the public address as a parameter. So, two separate questions:

- Where exactly is the public address stored?
- Why cannot the spi driver access that public address directly instead of passing it to it, given that the HCI driver lives in the same address space of the application and host.

arbrauns commented:
The MAC address is stored in an application-specific config system (redundant paritions, upgrade/downgrade handling, that kind of thing). It's not stored at a specific (flash) memory address, it's loaded from the config at boot time.

At least for the BlueNRG-MS, the address *can't* be stored on the controller, since it doesn't have non-volatile memory for it.

msmttchr commented:
> > The MAC address is stored in an application-specific config system (redundant paritions, upgrade/downgrade handling, that kind of thing). It's not stored at a specific (flash) memory address, it's loaded from the config at boot time.
> 
> OK, but why cannot the BlueNRG driver access this config system directly? is there something preventing the driver from doing it? It just feels a bit like overkill to create a new path from Host to HCI driver just to provide the public address, which typically never comes from the application.

BlueNRG-MS is a controller only device when used in Zephyr and the external microcontroller connected via SPI has the burden to set the public address, if used. See also https://www.st.com/resource/en/design_tip/dt0070-how-to-set-the-bluetooth-device-address-on-bluenrgms-stmicroelectronics.pdf

arbrauns commented:
> OK, but why cannot the BlueNRG driver access this config system directly? is there something preventing the driver from doing it?

Please elaborate, surely you can't be suggesting that I hard code our bespoke, application-specific config system into an upstream Zephyr driver?

> It just feels a bit like overkill to create a new path from Host to HCI driver just to provide the public address, which typically never comes from the application.

As mentioned, for BlueNRG, the public address *always* comes from the application, simply because there's nowhere else it could come from.

alwa-nordic commented:
> > OK, but why cannot the BlueNRG driver access this config system directly? is there something preventing the driver from doing it?
> 
> Please elaborate, surely you can't be suggesting that I hard code our bespoke, application-specific config system into an upstream Zephyr driver?

Would it work for you to upload the address to the controller in the `bt_hci_transport_setup` hook (defined in `hci_driver.h`)? Uploading the address is analogous to uploading firmware.

jhedberg commented:
> Would it work for you to upload the address to the controller in the `bt_hci_transport_setup` hook (defined in `hci_driver.h`)? Uploading the address is analogous to uploading firmware.

`bt_hci_transport_setup` is a call done from within the `open()` call, i.e. the mechanisms & flow control for sending HCI commands isn't initialised yet at this point. We have the `hci_driver->setup()` callback for this reason (for things that cannot be done inside `open()`). Btw, I realise the unfortunate naming with both of these using "setup" in their names, but `transport_setup()` is really about the physical HCI transport, i.e. for low-level actions to enable communication over the transport.

jhedberg commented:
> > Would it work for you to upload the address to the controller in the `bt_hci_transport_setup` hook (defined in `hci_driver.h`)? Uploading the address is analogous to uploading firmware.
> 
> `bt_hci_transport_setup` is a call done from within the `open()` call, i.e. the mechanisms & flow control for sending HCI commands isn't initialised yet at this point. We have the `hci_driver->setup()` callback for this reason (for things that cannot be done inside `open()`). Btw, I realise the unfortunate naming with both of these using "setup" in their names, but `transport_setup()` is really about the physical HCI transport, i.e. for low-level actions to enable communication over the transport.

One other point about `open()` vs `setup()`: the latter is only ever run once, whereas `open()` can be called multiple times if `close()` is also supported. I.e. you can get a sequence of something like: `open()`, `setup()`, `close()`, `open()`, `close()`, `open()`, and so on.

carlescufi commented:
> > OK, but why cannot the BlueNRG driver access this config system directly? is there something preventing the driver from doing it?
> 
> Please elaborate, surely you can't be suggesting that I hard code our bespoke, application-specific config system into an upstream Zephyr driver?

No no, not at all. I though that you had an STM BlueNRG-specific config system in your image that the BlueNRG controller understood. In that case the solution I was suggesting is not feasible.

Coincidentally, where are you getting the public address from to then store it in your config system? STM themselves?

arbrauns commented:
> Coincidentally, where are you getting the public address from, so you can then store it in your config system? STM themselves?

It is taken from the product manufacturer's own, IEEE-allocated MAC address space. It is saved into the config system in the factory.

carlescufi reviewed (changes requested):
Hi @arbrauns, sorry for the delay. We discussed this in today's [Bluetooth meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-bluetooth-meeting) and we came to the conclusion that it'd be cleaner to take a slightly different approach.

Instead of adding a new `set_public_addr` method to the HCI driver struct, we'd rather pass a parameter to the existing `setup` call. This would look like this:

```
int (*setup)(struct bt_hci_setup_params* params);

struct bt_hci_setup_params {
#ifdef CONFIG_BT_HCI_SET_PUBLIC_ADDR
    bt_addr_t *addr;
#endif
};
```

Technically this is a breaking API change. We can discuss if this really needs to go through the [Stable API change procedure](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes) given how unusual is to have out-of-tree HCI drivers.
It certainly requires a mention in the migration guide.

CC @cvinayak @jhedberg @alwa-nordic @jori-nordic @Thalley

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3638,9 +3638,23 @@ static void hci_vs_init(void)
+
+#if defined(CONFIG_BT_HCI_SETUP)
+	struct bt_hci_setup_params setup_params = { 0 };
+#endif
```
Can't you just remove this `#endif` and let everything be inside the same `#if defined (CONFIG_BT_HCI_SETUP)` block? As for the "Setting public address not supported by driver" error, wouldn't it be better to give that already when calling `bt_id_create()` rather than here?

carlescufi reviewed (approved):
LGTM, thank you!

msmttchr commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3638,9 +3638,18 @@ static void hci_vs_init(void)
+
 	if (bt_dev.drv->setup) {
-		err = bt_dev.drv->setup();
+		err = bt_dev.drv->setup(&setup_params);
```
@arbrauns, I think this code is executed during initialization, but if the host send an HCI_RESET, this setting is lost in BlueNRG. I am not sure whether you already took into account this case (see also `bt_spi_send`, to check how HCI_RESET is managed).

arbrauns replied:
I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?

jhedberg replied:
> I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?

That sounds like a potential solution to me (though I think the question was directed at @msmttchr ).

Another option is to set `hci_driver.quirks = BT_QUIRK_NO_RESET` - that will cause the host to never send the `HCI_Reset` command.

msmttchr replied:
> I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?

Well, I thought you could use read_config prior to host reset to save public address and restore it after reset is complete, but I also seen @jhedberg comment on BT_QUIRK_NO_RESET, so I am no longer sure this extra complication is beneficial.

msmttchr replied:
> > I had taken this into account initially (setting the public address in `bt_spi_handle_vendor_evt`), but it was lost with the switch to generic HCI API functions. I guess I could still save the address in `bt_spi_bluenrg_setup` and then restore it in `bt_spi_handle_vendor_evt`?
> 
> 
> 
> That sounds like a potential solution to me (though I think the question was directed at @msmttchr ).
> 
> 
> 
> Another option is to set `hci_driver.quirks = BT_QUIRK_NO_RESET` - that will cause the host to never send the `HCI_Reset` command.

Ah ah, I think this is a better proposal to keep things reasonably linear. I initially thought that this setting prevented reset only during initialization, but if this is meant to prevent sending HCI reset in all cases, I think this is just fine.

msmttchr reviewed (commented):
I left a comment for HCI_RESET behaviour, for the rest LGTM.

msmttchr reviewed (approved):
Ok for me. Thanks

jori-nordic commented on include/zephyr/drivers/bluetooth/hci_driver.h:
```diff
@@ -139,6 +139,14 @@ enum bt_hci_driver_bus {
+#if defined(CONFIG_BT_HCI_SETUP) || defined(__DOXYGEN__)
+struct bt_hci_setup_params {
+#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
+	bt_addr_t *public_addr;
```
please add some documentation on what this contains, and when it contains it.

jori-nordic commented on include/zephyr/drivers/bluetooth/hci_driver.h:
```diff
@@ -139,6 +139,14 @@ enum bt_hci_driver_bus {
+#if defined(CONFIG_BT_HCI_SETUP) || defined(__DOXYGEN__)
+struct bt_hci_setup_params {
+#if defined(CONFIG_BT_HCI_SET_PUBLIC_ADDR)
+	bt_addr_t *public_addr;
```
also why is it a pointer? It would probably be simpler to explicitly copy the address into the params.

arbrauns replied:
Sure, it's possible, but making it a pointer gives a convenient way of representing "no public address" as `NULL`.

jori-nordic commented on tests/bluetooth/host/id/bt_id_create/src/test_suite_invalid_inputs.c:
```diff
@@ -78,26 +78,6 @@ ZTEST(bt_id_create_invalid_inputs, test_id_list_is_full)
- *  Expected behaviour:
- *   - '-EINVAL' error code is returned representing invalid values were used.
- */
-ZTEST(bt_id_create_invalid_inputs, test_public_address)
```
this behavior is still correct if `CONFIG_BT_HCI_SET_PUBLIC_ADDR=n`.
Please don't remove the test.

arbrauns replied:
Added a `ztest_tesst_skip()` instead.

jori-nordic commented on tests/bluetooth/host/id/bt_id_create/src/main.c:
```diff
@@ -215,3 +215,36 @@ ZTEST(bt_id_create, test_create_id_valid_input_address)
+ *   - The public address is loaded to bt_dev.id_addr[BT_ID_DEFAULT]
+ *   - bt_dev.id_count is incremented
+ */
+ZTEST(bt_id_create, test_public_address)
```
oh no
![image](https://github.com/zephyrproject-rtos/zephyr/assets/68428270/da6785fc-8a81-44db-878e-2600e89d5a9b)

See this branch for how to fix the test.
https://github.com/jori-nordic/zephyr/tree/49f1414d9dc74eb2fc3e23fc6ec4d23efffb4108/./

jori-nordic replied:
edit: commit in question https://github.com/jori-nordic/zephyr/commit/cc2c1611654bcf77a7490418b4dcfdc4ae4a3341

arbrauns replied:
Thanks for taking care of that! I was wondering how to fix those tests.

jori-nordic commented on subsys/bluetooth/host/id.c:
```diff
@@ -1281,20 +1281,27 @@ int bt_id_create(bt_addr_le_t *addr, uint8_t *irk)
 {
```
You're changing behavior.
Please update [id_create's](https://github.com/jori-nordic/zephyr/blob/49f1414d9dc74eb2fc3e23fc6ec4d23efffb4108/include/zephyr/bluetooth/bluetooth.h#L359) documentation.

jori-nordic commented on include/zephyr/drivers/bluetooth/hci_driver.h:
```diff
@@ -139,6 +139,14 @@ enum bt_hci_driver_bus {
 	BT_HCI_DRIVER_BUS_IPM           = 9,
 };
 
+#if defined(CONFIG_BT_HCI_SETUP) || defined(__DOXYGEN__)
```
This can in some cases result in an empty struct. 
It compiles on my GCC, but might not on other toolchains.

arbrauns replied:
Any ideas how best to fix this? Add a dummy field, or always include `public_addr`?

jori-nordic replied:
I would vote for always include until someone complains

arbrauns commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -52,6 +52,11 @@ Device Drivers and Device Tree
+  ``CONFIG_BT_HCI_SETUP``) has gained a function parameter of type ``struct bt_hci_setup_params *``.
+  By default, the struct is empty, but drivers can opt-in to ``CONFIG_BT_HCI_SET_PUBLIC_ADDR`` if
+  they support setting the controller's public identity address, which will then be passed in the
+  ``pub_addr`` field.
```
Please check if it's okay now, I haven't been able to find any documentation on which rst markup syntaxes are available.

arbrauns commented:
Addressed review comments, and I've also changed the `struct bt_hci_setup_params *` parameter in `setup()` to `const struct bt_hci_setup_params *`.

jhedberg commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -224,23 +224,42 @@ static bool exit_irq_high_loop(void)
+int bt_spi_bluenrg_setup(const struct bt_hci_setup_params *params)
+{
+	int ret;
+	const bt_addr_t addr = params->public_addr;
```
Why not `const bt_addr_t *addr = &params->public_addr;`? Would be a bit more efficient than making a copy of the entire struct. Anyway, a rather minor issue. I won't block the PR for that.

msmttchr commented:
@arbrauns, after this was merged, I have noticed that setting the public address seems mandatory for platform using BlueNRG-MS. I am not sure, if this was intended or I miss something, but simply executing a beacon application will fail with `-EINVAL` error returned by `hci_init`.

See code fragment:

https://github.com/zephyrproject-rtos/zephyr/blob/76791cd708279692d198e03c51c8b2ff559fa7e5/subsys/bluetooth/host/hci_core.c#L3672-L3678

Root cause seems that setting `BT_BLUENRG_ACI` will force `BT_HCI_SET_PUBLIC_ADDR` and `BT_HCI_SETUP` to be set.

Any hint or comment from you side ?

jori-nordic commented:
Yeah it seems [this line](https://github.com/zephyrproject-rtos/zephyr/blob/76791cd708279692d198e03c51c8b2ff559fa7e5/subsys/bluetooth/host/hci_core.c#L3684) combined with [this one](https://github.com/zephyrproject-rtos/zephyr/blob/76791cd708279692d198e03c51c8b2ff559fa7e5/drivers/bluetooth/hci/spi.c#L252) is the issue.

I would advise for a simple `LOG_WRN("No public address is set");` and `return 0` instead.

arbrauns commented:
> Yeah it seems [this line](https://github.com/zephyrproject-rtos/zephyr/blob/76791cd708279692d198e03c51c8b2ff559fa7e5/subsys/bluetooth/host/hci_core.c#L3684) combined with [this one](https://github.com/zephyrproject-rtos/zephyr/blob/76791cd708279692d198e03c51c8b2ff559fa7e5/drivers/bluetooth/hci/spi.c#L252) is the issue.

Oh, yeah, that second one is a leftover from when this was a dedicated function rather than an optional argument to `setup()`, I didn't clean that up properly.

> I would advise for a simple `LOG_WRN("No public address is set");` and `return 0` instead.

Honestly doesn't even need the `LOG_WRN()`, if the user didn't supply a public address, that's probably intentional. The whole `bt_spi_send_aci_config()` should just be gated by `if (!bt_addr_eq(&addr, BT_ADDR_NONE) && !bt_addr_eq(&addr, BT_ADDR_ANY))`.

HoZHel commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -120,6 +120,7 @@ config BT_SPI_INIT_PRIORITY
 
 config BT_BLUENRG_ACI
 	bool "ACI message with with BlueNRG-based devices"
+	select BT_HCI_SET_PUBLIC_ADDR
```
I suggest you to remove [this line](https://github.com/arbrauns/zephyr/blob/11fcb16620c30f6c2ec9033213e590d07dc99fd9/drivers/bluetooth/hci/Kconfig#L123) as it forces BlueNRG devices to use public address. Instead, in the definition of `BT_HCI_SET_PUBLIC_ADDR` you can use `depends on BT_BLUENRG_ACI` to make it optional.

arbrauns replied:
That doesn't sound right, `BT_HCI_SET_PUBLIC_ADDR` should just not be "forcing to use a public address" in the first place. It should only enable support for setting a public address if desired.

`depends on BT_BLUENRG_ACI` isn't generally correct either, the functionality could be implemented by other drivers as well.

msmttchr replied:
Well, I guess other drivers can be added to the depend list and with PR #65988, you are removing the concept of "forcing" which unintentionally, present in this PR.

jhedberg replied:
`select` is correct here. It's indicating *support* for a specific feature, not that the feature must be used.

arbrauns commented:
I've opened #65988.
